### ISSUE TITLE
CVE-2024-7592: Denial of Service Vulnerability in `http.cookies._unquote()`

### ISSUE THREAD

**Issue by ch4n3-yoon** [python/cpython#123067]
# Bug report

## Description

A potential Denial of Service (DoS) vulnerability, identified as CVE-2024-7592, has been discovered in the `_unquote()` method of the `http.cookies` module in Python's standard library. This vulnerability is particularly concerning as it affects frameworks that utilize this method, including Django.

### Vulnerable Code

The `_unquote()` function uses regular expressions `_OctalPatt` and `_QuotePatt` within a while loop to process input strings. The problematic patterns and their application can lead to exponential time complexity under certain conditions, akin to a Regular Expression Denial of Service (ReDoS) attack.

```python
# http/cookies.py
_OctalPatt = re.compile(r"\\[0-3][0-7][0-7]")
_QuotePatt = re.compile(r"[\\].")
def _unquote(str):
    # ... (code omitted for brevity)
    while 0 <= i < n:
        o_match = _OctalPatt.search(str, i)
        q_match = _QuotePatt.search(str, i)
        # ... (further processing)
```

## Impact
This vulnerability has also been verified in the Django framework, where the `parse_cookie()` function uses this method to process incoming cookie headers. This could potentially be exploited by sending specially crafted cookie values to trigger significant delays:

- Cookie sizes of 8000+ bytes caused delays of approximately 0.15 seconds per HTTP request.
- Cookie sizes of 20000+ bytes resulted in delays of about 1 second per request.

While many environments limit HTTP request sizes, the specific limits vary, and in some cases, this vulnerability could be exploited.

### CPython versions tested on:

CPython main branch

### Operating systems tested on:

Linux

### Linked PRs
* gh-123066
* gh-123075
* gh-123103
* gh-123104
* gh-123105
* gh-123106
* gh-123107
* gh-123108

**Comment by serhiy-storchaka** [comment]
The complexity is quadratic, not exponential.

**Comment by sethmlarson** [comment]
Closed in https://github.com/python/cpython/pull/123075

**Comment by hauntsaninja** [comment]
@sethmlarson looks like the backports are all unmerged (usually we keep the issue open till that's done)
Do you know if there's anything we're waiting on to merge them? If not, I can at least merge to 3.13 and 3.12! :-)

**Comment by sethmlarson** [comment]
@hauntsaninja I don't think there's anything blocking the backports!

**Comment by hauntsaninja** [comment]
Great, I merged to 3.13 and 3.12, the other branches will need the RM to merge

**Comment by serhiy-storchaka** [comment]
AFAIK backports to 3.13 should be backported by other core developer. This is why I did not merge them myself.
